core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java

```diff
@@ -629,7 +630,9 @@ public PlanOptimizers(
                         costCalculator,
                         // Temporary hack: separate optimizer step to avoid the sample node being replaced by filter before pushing
                         // it to table scan node
-                        ImmutableSet.of(new ImplementBernoulliSampleAsFilter(metadata))),
+                        ImmutableSet.of(
+                                new ImplementBernoulliSampleAsFilter(metadata),
+                                new ImplementTableFunctionSource(metadata))), // must run after RewriteTableFunctionToTableScan
```

Comment:
This is because the old rule `RewriteTableFunctionToTableScan` is applied to `TableFunctionNode`. If the node gets rewritten to `TableFunctionProcessingNode`, we can no longer push down.

Reply:
Add that explanation to the comment